system-parachains/coretime/coretime-kusama/src/coretime.rs

```diff
@@ -0,0 +1,280 @@
+			Ok(_) => log::debug!(
+				target: "runtime::coretime",
+				"Instruction to credit account sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Instruction to credit account failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn assign_core(
+		core: CoreIndex,
+		begin: RCBlockNumberOf<Self>,
+		assignment: Vec<(CoreAssignment, PartsOf57600)>,
+		end_hint: Option<RCBlockNumberOf<Self>>,
+	) {
+		use crate::coretime::CoretimeProviderCalls::AssignCore;
+		let assign_core_call =
+			RelayRuntimePallets::Coretime(AssignCore(core, begin, assignment, end_hint));
+
+		// Weight for `assign_core` from Kusama runtime benchmarks:
+		// `ref_time` = 10177115 + (1 * 25000000) + (2 * 100000000) + (57600 * 13932) = 937660315
+		// `proof_size` = 3612
+		// Add 5% to each component and round to 2 significant figures.
+		let call_weight = Weight::from_parts(980_000_000, 3800);
+
+		let message = Xcm(vec![
+			Instruction::UnpaidExecution {
+				weight_limit: WeightLimit::Unlimited,
+				check_origin: None,
+			},
+			Instruction::Transact {
+				origin_kind: OriginKind::Native,
+				require_weight_at_most: call_weight,
+				call: assign_core_call.encode().into(),
+			},
+		]);
+
+		match PolkadotXcm::send_xcm(Here, Location::parent(), message.clone()) {
+			Ok(_) => log::debug!(
+				target: "runtime::coretime",
+				"Core assignment sent successfully."
+			),
+			Err(e) => log::error!(
+				target: "runtime::coretime",
+				"Core assignment failed to send: {:?}",
+				e
+			),
+		}
+	}
+
+	fn check_notify_revenue_info() -> Option<(RCBlockNumberOf<Self>, Self::Balance)> {
+		let revenue = CoretimeRevenue::get();
+		CoretimeRevenue::set(&None);
+		revenue
+	}
+
+	#[cfg(feature = "runtime-benchmarks")]
+	fn ensure_notify_revenue_info(when: RCBlockNumberOf<Self>, revenue: Self::Balance) {
+		CoretimeRevenue::set(&Some((when, revenue)));
+	}
+}
+
+/// Implements the [`AdaptPrice`] trait to control the price changes of bulk coretime. This tweaks
+/// the [`pallet_broker::Linear`] implementation which hard-corrects to 0 if cores are offered but
+/// not sold for just one sale. The monotonic lead-in behaviour is unchanged, while the change in
+/// base price between sales has a lower limit of 1/4 to match the upper limit of 4.
+pub struct LinearPlusC;
+impl AdaptPrice for LinearPlusC {
+	fn leadin_factor_at(when: FixedU64) -> FixedU64 {
+		// Start at 2x the base price and decrease to the base price at the end of leadin.
+		FixedU64::from(2).saturating_sub(when)
+	}
+
+	fn adapt_price(sold: CoreIndex, target: CoreIndex, limit: CoreIndex) -> FixedU64 {
+		if sold <= target {
+			// Range of [0.25, 1.0].
+			FixedU64::from_rational(1, 4)
+				.saturating_add(FixedU64::from_rational((3 * sold).into(), (4 * target).into()))
+		} else {
+			// Range of (1.0, 2.0].
+			FixedU64::one().saturating_add(FixedU64::from_rational(
+				(sold - target).into(),
+				(limit - target).into(),
```

Comment:
We are certain that `limit < target`? I mean it makes sense but especially as we are dividing by this difference, if it is zero then this could cause problems.

Reply:
In this branch of the `if` we know that sold < target.
The `limit` must be `>= sold` by construction, and thus `target` must be `> limit`.

If you were to use this anywhere else except as a PriceAdapter for the broker it wouldn't be certain I suppose, but it's a pretty tightly coupled interface.